rust/src/mask/mask_object/serialization.rs

```diff
@@ -0,0 +1,130 @@
+use crate::{
+    mask::{config::serialization::MASK_CONFIG_BUFFER_LEN, MaskConfig},
+    message::{utils::range, DecodeError, FromBytes, ToBytes},
+};
+
+const MASK_CONFIG_FIELD: Range<usize> = range(0, MASK_CONFIG_BUFFER_LEN);
+const DIGITS_FIELD: Range<usize> = range(MASK_CONFIG_FIELD.end, 4);
+
+struct MaskObjectBuffer<T> {
+    inner: T,
+}
+impl<T: AsRef<[u8]>> MaskObjectBuffer<T> {
+    pub fn new(bytes: T) -> Result<Self, DecodeError> {
+        let buffer = Self { inner: bytes };
+        buffer
+            .check_buffer_length()
+            .context("not a valid MaskObject")?;
+        Ok(buffer)
+    }
+
+    pub fn new_unchecked(bytes: T) -> Self {
+        Self { inner: bytes }
+    }
+
+    pub fn check_buffer_length(&self) -> Result<(), DecodeError> {
+        let len = self.inner.as_ref().len();
+        if len < DIGITS_FIELD.end {
+            return Err(anyhow!(
+                "invalid buffer length: {} < {}",
+                len,
+                DIGITS_FIELD.end
+            ));
+        }
+
+        let config = MaskConfig::from_bytes(&self.config()).context("invalid MaskObject buffer")?;
+        let bytes_per_digit = config.bytes_per_digit();
+        let (data_length, overflows) = (self.digits() as usize).overflowing_mul(bytes_per_digit);
+        if overflows {
+            return Err(anyhow!(
+                "invalid MaskObject buffer: invalid mask config or digits field"
+            ));
+        }
+        let total_expected_length = DIGITS_FIELD.end + data_length;
+        if len < total_expected_length {
+            return Err(anyhow!(
+                "invalid buffer length: expected {} bytes but buffer has only {} bytes",
+                total_expected_length,
+                len
+            ));
+        }
+        Ok(())
+    }
+
+    pub fn digits(&self) -> u32 {
+        // UNWRAP SAFE: the slice is exactly 4 bytes long
+        u32::from_be_bytes(self.inner.as_ref()[DIGITS_FIELD].try_into().unwrap())
+    }
+
+    pub fn config(&self) -> &[u8] {
+        &self.inner.as_ref()[MASK_CONFIG_FIELD]
+    }
+
+    pub fn data(&self) -> &[u8] {
+        &self.inner.as_ref()[DIGITS_FIELD.end..]
+    }
+}
+
+impl<T: AsMut<[u8]>> MaskObjectBuffer<T> {
+    pub fn set_digits(&mut self, value: u32) {
+        self.inner.as_mut()[DIGITS_FIELD].copy_from_slice(&value.to_be_bytes());
+    }
+    pub fn config_mut(&mut self) -> &mut [u8] {
+        &mut self.inner.as_mut()[MASK_CONFIG_FIELD]
+    }
+
+    pub fn data_mut(&mut self) -> &mut [u8] {
+        &mut self.inner.as_mut()[DIGITS_FIELD.end..]
+    }
+}
+
+impl ToBytes for MaskObject {
+    fn buffer_length(&self) -> usize {
+        DIGITS_FIELD.end + self.config.bytes_per_digit() * self.data.len()
+    }
+
+    fn to_bytes<T: AsMut<[u8]>>(&self, buffer: &mut T) {
+        let mut writer = MaskObjectBuffer::new_unchecked(buffer.as_mut());
+        self.config.to_bytes(&mut writer.config_mut());
+        writer.set_digits(self.data.len() as u32);
```

Comment:
yes by storing the length on 4 bytes it is implicitely assumed that we can't have more than 2^32 weights. Is this too much of a restriction? I can make this an u64 instead.

Reply:
i'm not sure, thought about this before as well. for now it is probably more than enough, but i'm tending towards failing loudly (for example with a `panic!`) instead of silently, such that this is not a reason for bugs in the far future.

Reply:
Good idea I'll do that.